### PR TITLE
Support on disk GeoWebCache tilecaches

### PR DESCRIPTION
**Pull request by bjornharrtell**
This PR provides a Source that can read tiles directly from a preseeded GeoWebCache cache dir that has been exposed on a webserver.

Adapted from @tschaub's experimental code for OpenLayers 2 here:
http://dev.openlayers.org/sandbox/tschaub/gwc/lib/OpenLayers/Layer/GeoWebCache.js

### CODE REVIEW

**Comment by sweco-sebhar** [comment]
I think this is ready for review.

**Comment by elemoine** [comment]
Note that the build fails on Travis.

**Comment by bjornharrtell** [comment]
Fixed now.

**Comment by elemoine** [comment]
LGTM. Does anyone from Boundless interested in reviewing this? Are preseeded GWC caches common and actually used in practice? I have no opinion, just asking.

**Comment by bartvde** [comment]
I think this is useful

@elemoine so you have reviewed the code already? If so, I think we should merge this in?

File: externs/olx.js [hunk 1]
```diff
@@ -5556,6 +5556,136 @@ olx.source.WMTSOptions.prototype.urls;
...
+/**
+ * URL to GeoWebCache base storage catalog
+ * @type {string}
+ * @api stable
+ */
+olx.source.GWCOptions.prototype.url;
+
+
+/**
+ * Layer name
```
Comment (elemoine) [inline externs/olx.js]: Please add a dot at end of line. We use punctuation elsewhere.

File: externs/olx.js [hunk 2]
```diff
@@ -5556,6 +5556,136 @@ olx.source.WMTSOptions.prototype.urls;
...
+ * @type {string}
+ * @api stable
+ */
+olx.source.GWCOptions.prototype.layer;
+
+
+/**
+ * Gridset name
```
Comment (elemoine) [inline externs/olx.js]: Punctuation. And doesn't this deserve a better description?

File: src/ol/source/gwcsource.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    logo: options.logo,
+    projection: projection,
+    tileGrid: tileGrid,
+    tileLoadFunction: options.tileLoadFunction,
+    tilePixelRatio: options.tilePixelRatio,
+    tileUrlFunction: this.tileUrlFunction
+  });
+
+  /**
+   * @type {string}
```
Comment (elemoine) [inline src/ol/source/gwcsource.js]: Could `url`,  `layer` and `gridset` be private?

**Comment by elemoine** [comment]
@bartvde I added comments, but I think I have more.

**Comment by bartvde** [comment]
@elemoine cool TIA it wasn't clear to me if you had already reviewed (since you said LGTM

**Comment by elemoine** [comment]
I had looked at it, but another look actually raised comments  :-)

**Comment by bjornharrtell** [comment]
Thanks for the double review @elemoine :) Should be amended now.

**Comment by bjornharrtell** [comment]
Hang on, gjslint didn't like my @private additions. I'll try to fix that and update.

**Comment by elemoine** [comment]
My last comment would be about support for multi-domain URLs, which other tile image sources support.

**Comment by bjornharrtell** [comment]
Looking at how other tiled sources works with URLs it might be possible to rewrite this PR to use ol.TileUrlFunction.createFromTemplate and friends and in the process get support for multi-domain URLs.

**Comment by elemoine** [comment]
Exactly.

**Comment by ahocevar** [comment]
Closing due to lack of activity. Also, the way we create tile urls has changed since. @bjornharrtell, please reopen if you are able/willing to continue working on this. And sorry for letting your valuable work go unnoticed for such a long time.

**Pull request closed without merging**